transaction/builder/src/transaction_builder.rs

```diff
@@ -447,6 +447,56 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
         amount: Amount,
         recipient: &PublicAddress,
         rng: &mut RNG,
+    ) -> Result<TxOutContext, TxBuilderError> {
+        self.add_output_with_tx_private_key(amount, recipient, None, rng)
+    }
+
+    /// Add a non-change output to the transaction, optionally with a specified
+    /// tx_private_key.
+    ///
+    /// Specifying the tx_private_key gives you two things:
+    ///
+    /// * Together with the amount and recipient, fixes the generated TxOut
+    ///   public key, target key, and blinding factor for amount. Because the
+    ///   blockchain enforces that TxOut public keys are unique, this is a point
+    ///   of mutual exclusion, and you can use this to create idempotent payment
+    ///   interfaces.
+    /// * If you know the tx private key, you can prove to an untrusting third
+    ///   party what the amount and recipient of the tx out is. You can use this
+    ///   resolve disputes. The TxOut shared secret and confirmation numbers
+    ///   don't accomplish this because they don't reveal the recipient.
+    ///
+    /// If the tx_private_key is not pseudorandom, it will harm the privacy of
+    /// transactions. For a merchant or exchange, a reasonable way to derive
+    /// it is to hash the payment id, or the withdrawal id, together with a
+    /// 32 byte secret. (You could use your private spend key or similar for
+    /// example, but you may wish to rotate that from time to time, and you
+    /// may not expect that idempotence would break across that key rotation.
+    /// It's up to the application developer to decide the most suitable
+    /// scheme.) You must ensure 32 bytes of pseudo-entropy to avoid
+    /// undermining the transaction protocol.
+    ///
+    /// An alternative is to seed the RNG that is used with the transaction
+    /// builder and then call add_output as usual. However, this approach
+    /// means that upgrading your RNG is a way that idempotence can break,
+    /// and in many cases, breaking idempotence means risk of double payments /
+    /// loss of funds. Setting the tx_private_key directly is possibly
+    /// simpler and with fewer hazards. Additionally, this approach allows
+    /// you to easily determine and record the tx_private_key, which you may
+    /// want for other reasons as described.
+    ///
+    /// # Arguments
+    /// * `amount` - The amount of this output
+    /// * `recipient` - The recipient's public address
+    /// * `tx_private_key` - Optionally, a specific tx_private_key to use for
+    ///   this output.
+    /// * `rng` - RNG used to generate blinding for commitment
+    pub fn add_output_with_tx_private_key<RNG: CryptoRng + RngCore>(
```

Comment:
this is a new transaction builder API that allows adding an output with a specified tx private key